Doc/library/sqlite3.rst

```diff
@@ -2358,6 +2340,111 @@ can be found in the `SQLite URI documentation`_.
 .. _SQLite URI documentation: https://www.sqlite.org/uri.html
 
 
+.. _sqlite3-howto-row-factory:
+
+How to create and use row factories
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` represent fetched rows as :class:`tuples <tuple>`.
+If a :class:`!tuple` does not suit your needs,
+use the :class:`sqlite3.Row` class or a custom :attr:`~Connection.row_factory`.
+
+:class:`!Row` provides indexed and case-insensitive named access to columns,
+with low memory overhead and minimal performance impact.
+In order to use :class:`Row` as a row factory,
+simply assign it to the :attr:`Connection.row_factory` attribute:
+
+.. doctest::
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = sqlite3.Row
+
+Query results are now processed using :class:`Row`:
+
+.. doctest::
+
+   >>> res = con.execute("SELECT 'Earth' AS name, 6378 AS radius")
+   >>> row = res.fetchone()
+   >>> row.keys()
+   ['name', 'radius']
+   >>> row[0]  # Access by index.
+   'Earth'
+   >>> row["name"]    # Access by name.
+   'Earth'
+   >>> row["RADIUS"]  # Column names are case-insensitive.
+   6378
+
+It is also possible to assign row factories to cursors using
+:attr:`Cursor.row_factory`:
+
+.. doctest::
+
+   >>> cur = con.cursor()
+   >>> cur.row_factory == con.row_factory
+   True
+   >>> cur.row_factory = None  # Override cursor row factory.
+
+   # The cursor and the connection row factories now differ.
+   >>> cur.row_factory == con.row_factory
+   False
+   >>> cur.execute("SELECT 'Hello'").fetchone()
+   ('Hello',)
+
+   # The connection still uses sqlite3.Row, as set in the previous example.
+   >>> con.execute("SELECT 'Hello'").fetchone()
+   <sqlite3.Row object at ...>
+
+If more flexibility is needed, implement a custom row factory.
+Here's an example of one returning a :class:`dict`:
+
+.. doctest::
+
+   >>> def dict_factory(cursor, row):
+   ...     col_names = [col[0] for col in cursor.description]
+   ...     return {key: value for key, value in zip(col_names, row)}
+
+   >>> con = sqlite3.connect(":memory:")
+   >>> con.row_factory = dict_factory
+   >>> for row in con.execute("SELECT 1 AS a, 2 AS b"):
+   ...     print(row)
+   {'a': 1, 'b': 2}
+
+Here's an example of a factory that return a :class:`~collections.namedtuple`.
+
+.. testcode::
+
+   from collections import namedtuple
+
+   def _get_fields(cursor):
+        return (col[0] for col in cursor.description)
+
+   def _make_cls(fields):
+       return namedtuple("Row", fields)
+
+   def namedtuple_factory(cursor, row):
+       cls = _make_cls(_get_fields(cursor))
```

Comment:
Since this boilerplate has little to do with the task at hand, and rather distracts and may make the reader wonder if this structure has some significance to `sqlite3`, why not simplify this to just:

```suggestion
   def namedtuple_factory(cursor, row):
       cls = namedtuple("Row", _get_fields(cursor))
```

Reply:
It does make describing the suggested optimization below less trivial, but as that is an optimization exercise for the reader and orthogonal to the `sqlite3` itself, my opinion would be that its not worth the complexity and confusion cost of adding it here just to make the optimization easier. Something something something premature optimization is the root of all eval...

Reply:
True! Let's keep it to the point.